# –ó–∞–≥—Ä—É–∑–∫–∞ –Ω–µ–æ–±—Ö–æ–¥–∏–º—ã—Ö –±–∏–±–ª–∏–æ—Ç–µ–∫ - –Ω–µ —É—Å—Ç–∞–Ω–∞–≤–æ–∏–≤–∞–µ–º 
library(dplyr)
library(tidyr)
library(SpiecEasi)
library(limma)
library(tibble) 

# –ú–Ω–æ–≥–æ—Å—Ç—Ä–æ—á–Ω—ã–π –∫–æ–º–º–µ–Ω—Ç–∞—Ä–∏–π
# # Command + Shift + C
# Ctrl + Shift + C

### =========================================================================================================================
# –ö–∞–Ω–æ–Ω–∏—á–µ—Å–∫–∏–π –∫–æ—Ä—Ä–µ—Å–ø–æ–Ω–¥–µ–Ω—Ç–Ω—ã–π –∞–Ω–∞–ª–∏–∑ (CCA)
# –û—Å–Ω–æ–≤–∞–Ω –Ω–∞ –∫–æ—Ä—Ä–µ—Å–ø–æ–Ω–¥–µ–Ω—Ç–Ω–æ–º –∞–Ω–∞–ª–∏–∑–µ, —Ç–æ –µ—Å—Ç—å –≤–º–µ—Å—Ç–æ –∏—Å—Ö–æ–¥–Ω—ã—Ö —Å—ã—Ä—ã—Ö, —Ü–µ–Ω—Ç—Ä–∏—Ä–æ–≤–∞–Ω–Ω—ã—Ö –¥–∞–Ω–Ω—ã—Ö (ùíÄ) –∏—Å–ø–æ–ª—å–∑—É–µ—Ç—Å—è –º–∞—Ç—Ä–∏—Ü–∞ ùë∏
# (–ú–∞—Ç—Ä–∏—Ü–∞ –≤–∫–ª–∞–¥–æ–≤ –≤ –æ—Ç–∫–ª–æ–Ω–µ–Ω–∏–µ –æ—Ç –Ω—É–ª–µ–≤–æ–π –º–æ–¥–µ–ª–∏, –æ—Å–Ω–æ–≤–∞–Ω–Ω–æ–π –Ω–∞ –Ω–µ–∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–∏ –º–µ–∂–¥—É –ø—Ä–∏–∑–Ω–∞–∫–∞–º–∏ –∏ –æ–±—ä–µ–∫—Ç–∞–º–∏).

 # - CCA - –∫–∞–Ω–æ–Ω–∏—á–µ—Å–∫–∏–π –∫–æ—Ä–µ—Å–ø–æ–Ω–¥–µ–Ω—Ç–Ω—ã–π –∞–Ω–∞–ª–∏–∑ –≤ vegan
 #   - –ó–∞–≤–∏—Å–∏–º—ã–µ –ø–µ—Ä–µ–º–µ–Ω–Ω—ã–µ (–æ—Ç–∫–ª–∏–∫–∏) - –æ–±–∏–ª–∏–µ –≤–∏–¥–æ–≤
 #   - –ù–µ–∑–∞–≤–∏—Å–∏–º—ã–µ –ø–µ—Ä–µ–º–µ–Ω–Ω—ã–µ (–ø—Ä–µ–¥–∏–∫—Ç–æ—Ä—ã) - —Ç—Ä–∞–Ω—Å–∫—Ä–∏–ø—Ç–æ–º ???
 #  


# –¢—Ä–∞–Ω—Å–ø–æ–Ω–∏—Ä–æ–≤–∞–Ω–∏–µ —Ç–∞–±–ª–∏—Ü –¥–ª—è —É–¥–æ–±—Å—Ç–≤–∞ –≤—ã—á–∏—Å–ª–µ–Ω–∏—è –∫–æ—Ä—Ä–µ–ª—è—Ü–∏–π

metagenome_raw <- metagenome_filltered %>% 
  
  group_by(Species) %>%
  filter(sum(Value_num == 0) / n() < 0.9) %>%
  ungroup() %>%
  
  pivot_wider(
    names_from = Species,
    values_from = Value_num,
    values_fill = 0  #  –ø—Ä–æ–ø—É—Å–∫–∏ –Ω—É–ª—è–º–∏ —Å—Ä–∞–∑—É –ø—Ä–∏ pivoting
  ) %>% 
  
  column_to_rownames(var = names(.)[1]) %>% 
  
  .[, colSums(. > 0) >= max(3, 0.25 * nrow(.))] %>% # –º–∏–Ω–∏–º—É–º 3  –æ–±—Ä–∞–∑—Ü–æ–≤ –≤–∏–¥–æ–≤ 25% –≤–∏–¥–æ–≤
  
  
  .[, colSums(.) >= 0.0005 * sum(.)] %>%
  
  mutate(across(everything(), ~replace_na(., 0))) %>% 
  apply(1, function(x) x / sum(x)) 

metagenome <- metagenome_raw %>% 
  decostand("hellinger") %>% 
  .[!rownames(.) %in% c("Mus Musculus", "Gallus gallus", "Meleagris gallopavo"), ] #–¥–æ–ª–æ–π —ç—É–∫–∞—Ä–∏–æ—Ç 




### =========================================================================================================================

# –£–¥–∞–ª—è–µ–º —Å—Ç—Ä–æ–∫–∏ (–æ—Ä–≥–∞–Ω–∏–∑–º—ã –∏–ª–∏ —Ç—Ä–∞–Ω—Å–∫—Ä–∏–ø—Ç—ã), –≥–¥–µ –±–æ–ª–µ–µ 90% –∑–Ω–∞—á–µ–Ω–∏–π —Ä–∞–≤–Ω—ã –Ω—É–ª—é
# –î–ª—è –º–µ—Ç–∞–≥–µ–Ω–æ–º–∞ - –≤—ã—à–µ

# –î–ª—è —Ç—Ä–∞–Ω—Å–∫—Ä–∏–ø—Ç–æ–º–∞ (–µ—Å–ª–∏ –¥–∞–Ω–Ω—ã–µ –Ω–µ –Ω–æ—Ä–º–∞–ª–∏–∑–æ–≤–∞–Ω—ã)
transcriptome <- transcriptome_matrix[rowSums(transcriptome_matrix == 0) / ncol(transcriptome_matrix) < 0.9, ]

# –ø–ª–∞–≤–∏–º —Ç–∞–±–ª–∏—Ü—É —Ç—Ä–∞—Å–Ω–∫—Ä–∏–ø—Ç–æ–º–∞ 


### =========================================================================================================================
# –Ω–æ—Ä–º–∞–ª–∏–∑–∞—Ü–∏—è —Å –ø–æ–º–æ—â—å—é –ø—Ä–æ–≥—Ä–∞–º–º—ã lima  


# transcriptome <- data.frame(lapply(transcriptome[, -1], function(x) as.numeric(as.character(x))))
# transcriptome <- sweep(transcriptome, 2, colSums(transcriptome), "/")


# –Ω–æ—Ä–º
metagenome <- sweep(metagenome, 2, colSums(metagenome, na.rm = TRUE), "/")
transcriptome <- sweep(transcriptome, 2, colSums(transcriptome), "/")


metagenome <- t(metagenome)
transcriptome <- t(transcriptome)


####### –ø–µ—Ä–µ–æ–ø—Ä–µ–¥–µ–ª–µ–Ω–∏–µ —Å—Ç—Ä–æ–∫ - —Å–æ–æ—Ç–≤–µ—Ç—Å–≤–∏–µ 

desired_order <- c("SRR8265549", "SRR8265548", "SRR8265622", "SRR8265550",
                   "SRR8265657", "SRR8265655", "SRR8265631", "SRR8265632",
                   "SRR8265651", "SRR8265658", "SRR8265652", "SRR8265594",
                   "SRR8265621", "SRR8265629")

transcriptome <- transcriptome[desired_order, ]
transcriptome <- transcriptome[!rownames(transcriptome) %in% "SRR8265631", ]



rownames(metagenome) <- c("SRR8265549", "SRR8265548", "SRR8265622", "SRR8265550",
                          "SRR8265657", "SRR8265655", "SRR8265632",
                          "SRR8265651", "SRR8265658", "SRR8265652", "SRR8265594",
                          "SRR8265621", "SRR8265629")

rownames(metagenome) == rownames(transcriptome)


# transcriptome <- data.frame(lapply(transcriptome, function(x) as.numeric(as.character(x))))

# T_reads.cca = cca(transcriptome~., metagenome, dist="bray", direction = 'forward', permutations = 999)


### =========================================================================================================================


# –ö–∞–∫ —Ä–∞–±–æ—Ç–∞–µ—Ç SPARCC?
# 
# - –õ–æ–≥–∞—Ä–∏—Ñ–º–∏—Ä–æ–≤–∞–Ω–∏–µ —Å–æ–æ—Ç–Ω–æ—à–µ–Ω–∏–π ‚Äî –≤–º–µ—Å—Ç–æ –∞–±—Å–æ–ª—é—Ç–Ω—ã—Ö –∑–Ω–∞—á–µ–Ω–∏–π –∏—Å–ø–æ–ª—å–∑—É—é—Ç—Å—è –ª–æ–≥–∞—Ä–∏—Ñ–º—ã –æ—Ç–Ω–æ—à–µ–Ω–∏–π –º–µ–∂–¥—É –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞–º–∏.
# 
# - –ò—Ç–µ—Ä–∞—Ç–∏–≤–Ω–∞—è –æ—Ü–µ–Ω–∫–∞ –∫–æ—Ä—Ä–µ–ª—è—Ü–∏–π ‚Äî –Ω–∞ –∫–∞–∂–¥–æ–º —à–∞–≥–µ –∏—Å–∫–ª—é—á–∞—é—Ç—Å—è —Å–ª–∞–±—ã–µ —Å–≤—è–∑–∏.
# 
# - –ë—É—Ç—Å—Ç—Ä–µ–ø-–ø—Ä–æ–≤–µ—Ä–∫–∞ ‚Äî –æ—Ü–µ–Ω–∫–∞ –∑–Ω–∞—á–∏–º–æ—Å—Ç–∏ –∫–æ—Ä—Ä–µ–ª—è—Ü–∏–π —á–µ—Ä–µ–∑ —Å–ª—É—á–∞–π–Ω—ã–µ –ø–æ–¥–≤—ã–±–æ—Ä–∫–∏.
# 

### =========================================================================================================================

# –§—É–Ω–∫—Ü–∏—è –¥–ª—è –≤—ã—á–∏—Å–ª–µ–Ω–∏—è –∫–æ—Ä—Ä–µ–ª—è—Ü–∏–π


calculate_correlations <- function(metagenome, transcriptome, method = "sparcc") {
  if (method == "sparcc") {
    # –ò—Å–ø–æ–ª—å–∑—É–µ–º SparCC –¥–ª—è –≤—ã—á–∏—Å–ª–µ–Ω–∏—è –∫–æ—Ä—Ä–µ–ª—è—Ü–∏–π
    sparcc_results <- SpiecEasi::sparcc(metagenome, transcriptome)
    correlations <- sparcc_results$cor
    p_values <- sparcc_results$pval
  } else {
    # –ò—Å–ø–æ–ª—å–∑—É–µ–º —Å—Ç–∞–Ω–¥–∞—Ä—Ç–Ω—É—é —Ñ—É–Ω–∫—Ü–∏—é cor
    correlations <- cor(metagenome, transcriptome, method = method)
    p_values <- matrix(NA, nrow = ncol(metagenome), ncol = ncol(transcriptome))  # p-–∑–Ω–∞—á–µ–Ω–∏—è –Ω–µ –≤—ã—á–∏—Å–ª—è—é—Ç—Å—è –≤ cor
  }
  return(list(correlations = correlations, p_values = p_values))
}

#corr_res <- calculate_correlations(metagenome_reduced, transcriptome_reduced)

# —Å–æ–ø–æ—Å—Ç–∞–≤–ª–µ–Ω–∏–µ - —É –Ω–∞—Å 154 —Ç–∞–∫—Å–æ–Ω–∞ - –≤—ã–≤–±–µ—Ä–µ–º –≥–µ–Ω—ã –ø–æ —ç–∫—Å–ø—Ä–µ—Å—Å–∏–∏

library(vegan) 

# –£–¥–∞–ª–µ–Ω–∏–µ –Ω–∏–∑–∫–æ—ç–∫—Å–ø—Ä–µ—Å—Å–∏—Ä—É–µ–º—ã—Ö –≥–µ–Ω–æ–≤
keep_genes <- rowSums(transcriptome > 0) >= 0.5*ncol(transcriptome) 

transcriptome_filt <- transcriptome[, keep_genes]

# 
# 
# –í—ã–±–æ—Ä —Ç–æ–ø-–≤–∞—Ä–∏–∞–±–µ–ª—å–Ω—ã—Ö –≥–µ–Ω–æ–≤ (–Ω–∞–ø—Ä–∏–º–µ—Ä, 500 —Å–∞–º—ã—Ö –∏–∑–º–µ–Ω—á–∏–≤—ã—Ö)
top_genes <- names(sort(apply(transcriptome_filt, 2, sd), decreasing = TRUE)[1:100])
# 
# 
# 

transcriptome_reduced <- transcriptome_filt[, top_genes] %>% 
  as.data.frame() %>%
  {rownames(.) <- rownames(transcriptome_filt); .} %>%  # –°–æ—Ö—Ä–∞–Ω—è–µ–º rownames
  apply(2, as.numeric) %>% 
  as.data.frame() %>%
  {rownames(.) <- rownames(transcriptome_filt); .}  


# correlations <- calculate_correlations(metagenome, transcriptome_reduced)

##################### CCA ###########################

reduction <- function(df, mincols = 3) {
  df %>%
    subset(apply(., 1, function(x) sum(x > 0, na.rm = TRUE) >= mincols)) # —É–±–∏—Ä–∞–µ–º –≤—ã—Ä–æ–∂–¥–µ–Ω–Ω—ã–µ –æ—Ä–≥–∞–Ω–∏–∑–º—ã
}

metagenome_reduced <- as.data.frame(reduction(metagenome))


metagenome_filtered <- metagenome_reduced %>%
  .[, colSums(. > 0) >= 0.10 * nrow(.)] %>% #  –ø–æ —Ä–∞—Å–ø—Ä–æ—Å—Ç—Ä–∞–Ω—ë–Ω–Ω–æ—Å—Ç–∏ (10% –æ–±—Ä–∞–∑—Ü–æ–≤)
  .[, colSums(.) >= 0.0001 * sum(.)]   #  –ø–æ abundance (0.01% –æ—Ç –æ–±—â–µ–≥–æ —á–∏—Å–ª–∞ reads)

# —á–∏—Å–ª–æ –Ω—É–ª–µ–π 

summary(colSums(metagenome_filtered == 0))

############################################################
# –ø–µ—Ä–≤—ã–π –≤–∞—Ä–∏–∞–Ω—Ç 

library(tidyverse)
library(ggrepel)
library(ggpubr)

# 1
cca_result <- cca(metagenome_filtered ~ ., data=transcriptome_reduced)

#data manipulation
sites <- scores(cca_result, display = "sites")
species <- scores(cca_result, display = "species")
env <- scores(cca_result, display = "bp")

df_sites <- as.data.frame(sites)
df_species <- as.data.frame(species) 

#add info
df_species <- df_species %>% 
  mutate(value = apply(metagenome_raw, 2, mean)[rownames(df_species)]) %>% 
  arrange(-value) %>% 
  mutate(label = NA)
df_species$label[1:10] <- rownames(df_species)[1:10]


df_env <- as.data.frame(env * 0.8) 

df_env <- df_env %>% 
  mutate(value = apply(transcriptome, 2, mean)[rownames(df_env)]) %>% 
  arrange(-value) %>% 
  rownames_to_column("label")

# plots

g1 <- ggplot() +
  geom_point(data=df_sites, aes(x=CCA1, y=CCA2), 
             color = "darkblue", size = 2, alpha = 0.7) +
  
  #geom_text_repel(data=df_sites, aes(x=CCA1, y=CCA2), label = rownames(df_sites), color = "steelblue", size = 3) +
  
  geom_point(data = df_species, aes(x = CCA1, y = CCA2, color = -log10(value)),
             shape = 17, size = 3) + 
  scale_color_gradientn("Metagenome,\nmean value, -lg", colors = c("lightgreen","pink", "purple"), n.breaks = 3)+
  ggnewscale::new_scale_color() +
  
  geom_segment(data = df_env, aes(x = 0, y = 0, xend = CCA1, yend = CCA2, color = -log10(value)),
               arrow = arrow(
                 length = unit(0.3, "cm"), 
                 angle = 20,  
                 type = "closed"
               )) +
  
  scale_color_gradientn("Transcriptome,\nmean value, -lg", colors = c("#F09590", "yellow","cadetblue4"), n.breaks = 3)+
  
  theme_minimal() +
  
  labs(x = paste0("CA1: ", round(cca_result$CCA$eig[1]/sum(cca_result$CCA$eig)*100, 1), "%"),
       y = paste0("CA2:", round(cca_result$CCA$eig[2]/sum(cca_result$CCA$eig)*100, 1), "%"),
       title = "CCA",
       subtitle = "Linkage of transcriptome and metagenome data") + 
  coord_fixed(ratio = 1) +
  theme(legend.position = "bottom")

g11 <- g1 + geom_text_repel(data = df_species, aes(x = CCA1, y = CCA2, label = label), 
                            color = c("#51107A"), fontface = "italic")
g12 <- g1 + geom_text_repel(data = df_env, aes(x = CCA1, y = CCA2, label = label),
                            color = "#115557", size = 4, fontface = "bold")

gg1 <- ggarrange(g11+xlab(""), g12 +labs(title = NULL, subtitle = NULL), ncol = 1, common.legend = T, legend = "bottom", align = "h")

# 2
cca_result <- cca(transcriptome_reduced ~ ., data=metagenome_filtered)

#data manipulation
sites <- scores(cca_result, display = "sites")
species <- scores(cca_result, display = "bp")
env <- scores(cca_result, display = "species")

df_sites <- as.data.frame(sites)
df_species <- as.data.frame(species) 

#add info
df_species <- df_species %>% 
  mutate(value = apply(metagenome_raw, 2, mean)[rownames(df_species) %>% str_remove_all("`")]) %>% 
  arrange(-value) %>% 
  rownames_to_column("label") %>% 
  mutate(label = label %>% str_remove_all("`"))

df_env <- as.data.frame(env * 0.8) 
df_env <- df_env %>% 
  mutate(value = apply(transcriptome, 2, mean)[rownames(df_env)]) %>% 
  arrange(-value) %>% 
  mutate(label = NA)

df_env$label[1:10] <- rownames(df_env)[1:10]

# plots

g2 <- ggplot() +
  geom_point(data=df_sites, aes(x=CCA1, y=CCA2), 
             color = "darkblue", size = 2, alpha = 0.7) +
  
  #geom_text_repel(data=df_sites, aes(x=CCA1, y=CCA2), label = rownames(df_sites), color = "steelblue", size = 3) +
  
  geom_point(data = df_env, aes(x = CCA1, y = CCA2, color = -log10(value)),
             shape = 17, size = 3) + 
  scale_color_gradientn("Transcriptome,\nmean value, -lg", colors = c("#F09590","yellow", "cadetblue4"), n.breaks = 3)+
  ggnewscale::new_scale_color() +
  
  geom_segment(data = df_species, aes(x = 0, y = 0, xend = CCA1, yend = CCA2, color = -log10(value)),
               arrow = arrow(
                 length = unit(0.3, "cm"), 
                 angle = 20,  
                 type = "closed"
               )) +
  
  
  scale_color_gradientn("Metagenome,\nmean value, -lg", colors = c("lightgreen","pink", "purple"), n.breaks = 3)+
  
  theme_minimal() +
  
  labs(x = paste0("CA1: ", round(cca_result$CCA$eig[1]/sum(cca_result$CCA$eig)*100, 1), "%"),
       y = paste0("CA2:", round(cca_result$CCA$eig[2]/sum(cca_result$CCA$eig)*100, 1), "%"),
       title = "CCA",
       subtitle = "Linkage of metagemome and transcriptome data") + 
  coord_fixed(ratio = 1) +
  theme(legend.position = "bottom")

g21 <- g2 + geom_text_repel(data = df_env, aes(x = CCA1, y = CCA2, label = label), 
                            color = c("#51107A"), fontface = "bold")
g22 <- g2 +  geom_text_repel(data = df_species, aes(x = CCA1, y = CCA2, label = label),
                             color = "#115557", size = 4, fontface = "italic") 

gg2 <- ggarrange(g21+xlab(""), g22 +labs(title = NULL, subtitle = NULL), ncol = 1, common.legend = T, legend = "bottom", align = "h")

ggpubr::ggarrange(gg1, gg2, align = "hv")


############################################################
# –∏–∑–≤–ª–µ—á–µ–Ω–∏–µ –Ω–∞–≥—Ä—É–∑–æ–∫ 

species_loadings <- scores(cca_result, display = "species", choices=1:2)
env_loadings <- scores(cca_result, display = "bp", choices=1:2)

write.csv(species_loadings, "species_loadings.csv")
write.csv(env_loadings, "env_loadings.csv")

############################################################
# –∫–æ—Ä—Ä–µ–ª—è—Ü–∏—è 


library(pheatmap)

loadings_matrix <- rbind(cca_result$CCA$v[, 1:2], 
                         cca_result$CCA$biplot[, 1:2])

rownames(loadings_matrix) <- c(colnames(metagenome_filtered), colnames(transcriptome_reduced))


pheatmap(loadings_matrix, 
         cluster_rows = FALSE, 
         cluster_cols = FALSE,
         display_numbers = TRUE,
         number_format = "%.2f",
         color = colorRampPalette(c("blue", "white", "brown"))(50),
         main = "Main HeatMap"
         )


correlations <- calculate_correlations(metagenome_filtered[, 1:80], transcriptome_reduced[, 1:80])


library(corrplot)
corrplot(correlations[1:10, 1:10])


vector_genera <- colnames(metagenome_reduced)



# write.table(metagenome_reduced, "metagenome_cca.tsv", sep="\t", row.names = TRUE)
# write.table(transcriptome_reduced, "transcriptome_cca.tsv", sep="\t", row.names = TRUE)

save.image(file = "CCA_environment.RData")

############################
# —Å—Ç–∞—Ä—ã–π –∫–æ–¥ –¥–ª—è –ø–æ—Å—Ç—Ä–æ–µ–Ω–∏—è –≥—Ä–∞—Ñ–∏–∫–∞ 

# (plot <- ggplot() +
#    geom_point(data=df_sites, aes(x=CCA1, y=CCA2), 
#               color = "steelblue", size = 3, alpha = 0.7) +
#    
#    geom_text_repel(data=df_sites, aes(x=CCA1, y=CCA2), label = rownames(df_sites), 
#                    color = "steelblue", size = 3) +
#    
#    geom_point(data = df_species, aes(x = CCA1, y = CCA2),
#               color = "firebrick", shape = 17, size = 3) + 
#    
#    geom_segment(data = df_env, aes(x = 0, y = 0, xend = CCA1, yend = CCA2),
#                 arrow = arrow(
#                   length = unit(0.3, "cm"), 
#                   angle = 20,              
#                 ), color = "purple") +
#    
#    geom_text_repel(data = df_env, aes(x = CCA1, y = CCA2, label = rownames(df_env)),
#                    color = "purple", size = 4, fontface = "bold") +
#    
#    theme_minimal() +
#    
#    labs(x = paste("CCA1 (", round(cca_result$CCA$eig[1]/sum(cca_result$CCA$eig)*100, 1), "%)"),
#         y = paste("CCA2 (", round(cca_result$CCA$eig[2]/sum(cca_result$CCA$eig)*100, 1), "%)"),
#         title = "CCA",
#         subtitle = "Linkage of transcriptome and metagemome data") + coord_fixed(ratio = 1))
# 

# cca_result <- cca(transcriptome_reduced ~ ., data=metagenome_filtered)
# summary(cca_result)
# 
# plot(cca_result)
# 
# # –æ—Ü–µ–Ω–∫–∞ –∑–Ω–∞—á–∏–º–æ—Å—Ç–∏ 
# 
# perm_test <- anova.cca(cca_result, permutations = 9999)
# print(perm_test)
# 
# 
# sites <- scores(cca_result, display = "sites")
# species <- scores(cca_result, display = "species")
# env <- scores(cca_result, display = "bp")
# 
# 
# # –≤ –¥–∞—Ç–∞—Ñ—Ä–µ–π–º
# df_sites <- as.data.frame(sites)
# df_species <- as.data.frame(species)
# df_env <- as.data.frame(env * 0.8) 
# 
# library(ggplot2)
# library(ggrepel) 


